### ISSUE TITLE
802.11s mesh packages cause DHCP to misbehave

### ISSUE THREAD

**Issue by etiago** [openNDS/mesh11sd#97]
Hi folks!

I'm neck deep into an issue that started happening when I upgraded to OpenWRT 23.05.5 (from, I think, 23.05.0). My device is a Totolink X5000R.

The issues originally started with my mesh, but I've brought everything down to a barebones configuration without a mesh as my setup was far too complicated to debug.

I now have a very basic OpenWRT installation where I'm running this device as an AP in an existing network. I have:
- DHCP is disabled.
- A static IP on this device.
- The wireless bridged with lan.
- One wireless SSID on each radio.

**The problem**
What I'm observing, and that is baffling me, is that when I have the following packages installed: `mesh11sd`, `kmod-nft-bridge`, `wpad-mbedtls` (or `wpad-openssl`, I've tried with each of them, and always with removing `wpad-basic-mbedtls`)

I sometimes (and it really is random) stop receiving responses to DHCP discovery requests. This happens both over the wifi and over the cable (with a cable connected to the AP).

This is purely with the packages present, but without any mesh configuration in place.

And then just by flashing a sysupgrade that only has `wpad-mbedtls` or just `wpad-openssl` everything starts working again and is rock solid for days/weeks.

The thing that's really confusing me is that this isn't a clear black/white, install the packages and everything stops working. Some devices are still able to get a DHCP allocation given enough time (a few long minutes) but clearly something is misbehaving as I've done tcpdumps and there are discoveries simply going unanswered.

**Thoughts**
What could be going on here?

All of this is in a device that I've done a factory reset on, so I'm sure there's no actual mesh11sd config. It's purely the presence of these packages that causes the issue.

It feels like a really basic setup where this device really should just be passing along the DHCP discoveries which should be trivial broadcast packets...

I've asked this question here: https://forum.openwrt.org/t/802-11s-mesh-packages-cause-dhcp-to-misbehave/214056 and @bluewavenet told me it'd be good to raise an issue here to investigate deeper.

What's a sensible next step here? Should I try installing just `mesh11sd` or just `kmod-nft-bridge` and see if the issue appears with just one of them installed?

And does anyone have any thoughts as to what the root cause might be?

**Comment by bluewavenet** [comment]
@etiago 

> Totolink X5000R

This device is a peculiar one. It has a "last generation" mt7621 SoC but has a "next generation" mt7915 wireless chip.

There were known issues with the mt79xx series chips but this was fixed prior to the release of 23.05.0. But as this device is unusual, perhaps it was not fixed on this one.... I don't know.

The issue was a buffer overflow in the wireless drives causing either a crash or for layer 3 traffic being dropped.

However this may not be anything to do with your current issue here.

Please, if you can, draw a diagram of your network so we can get an overview of the setup. 
Hint, for quickness, sketch it on a piece of paper and take a photo of it on your phone then post it here.

If you have an access point with an ethernet connection to your router, why do you need a mesh?

> I'm sure there's no actual mesh11sd config.

If mesh11sd is installed, then there will be a default config.

The defaults are:

1. enabled, with mesh11sd service running
2. manual mode, meaning it will use any existing mesh config in /etc/config/wireless, otherwise will sit waiting for a config.


Please show the outputs of the following:

`mesh11sd status`

`uci show mesh11sd`

`uci show network`

and

`uci show dhcp`

**Comment by etiago** [comment]
> The defaults are:
> 
> 1. enabled, with mesh11sd service running
> 2. manual mode, meaning it will use any existing mesh config in /etc/config/wireless, otherwise will sit waiting for a config.

Sure. I meant that I have no older mesh11sd config. It'll be with the defaults.


> Please, if you can, draw a diagram of your network so we can get an overview of the setup.
> Hint, for quickness, sketch it on a piece of paper and take a photo of it on your phone then post it here.

Sure thing. Here it is:
![photo_2024-11-10_05-34-56](https://github.com/user-attachments/assets/af6029cb-6917-4247-9900-df3caf64ca67)


> If you have an access point with an ethernet connection to your router, why do you need a mesh?

Because I have a few more Totolink devices which have no ethernet connections. And the network topology I'm trying to ultimately get to this this:
![photo_2024-11-10_05-35-00](https://github.com/user-attachments/assets/95cab71a-cf99-4588-a68f-56950ca8ea17)

Where a single AP has an ethernet connection to the Internet and acts as the mesh gateway, and the other mesh nodes connect to it.

Ultimately I think I don't necessarily need a mesh. I could have the other APs simply connect as clients to the one Totolink with the ethernet connection. But for some reason I wasn't able to get Client mode working previously and mesh was working fine before. Also, my understanding is that mesh is layer 2 whereas I think Client mode is layer 3... so mesh seemed a good fit for the design I'm aiming towards.

---

Outputs:

**mesh11sd status**

```
{
  "setup":{
    "version":"4.1.1",
    "enabled":"1",
    "procd_status":"running",
    "portal_detect":"1",
    "portal_detect_threshold":"0",
    "portal_channel":"default",
    "channel_tracking_checkinterval":"30",
    "mesh_basename":"m-11s-",
    "auto_config":"0",
    "auto_mesh_network":"lan",
    "auto_mesh_band":"2g40",
    "auto_mesh_id":"92d490daf46cfe534c56ddd669297e",
    "mesh_gate_enable":"1",
    "mesh_leechmode_enable":"0",
    "mesh_gate_encryption":"0",
    "txpower":"",
    "mesh_path_cost":"10",
    "mesh_path_stabilisation":"1",
    "checkinterval":"10",
    "interface_timeout":"10",
    "ssid_suffix_enable":"1",
    "debuglevel":"1"
  },
  "interfaces":{
  }
}
```

**uci show mesh11sd**

```
mesh11sd.setup=mesh11sd
mesh11sd.mesh_params=mesh11sd
mesh11sd.mesh_params.mesh_fwding='1'
mesh11sd.mesh_params.mesh_rssi_threshold='-65'
mesh11sd.mesh_params.mesh_gate_announcements='1'
mesh11sd.mesh_params.mesh_hwmp_rootmode='4'
mesh11sd.mesh_params.mesh_hwmp_rann_interval='5000'
mesh11sd.mesh_params.mesh_hwmp_root_interval='5000'
mesh11sd.mesh_params.mesh_hwmp_active_path_timeout='5000'
mesh11sd.mesh_params.mesh_hwmp_active_path_to_root_timeout='6000'
mesh11sd.mesh_params.mesh_max_peer_links='16'
mesh11sd.mesh_params.mesh_plink_timeout='10'
```

**uci show network**

```
network.loopback=interface
network.loopback.device='lo'
network.loopback.proto='static'
network.loopback.ipaddr='127.0.0.1'
network.loopback.netmask='255.0.0.0'
network.globals=globals
network.globals.ula_prefix='fd0a:9f65:4014::/48'
network.globals.packet_steering='1'
network.@device[0]=device
network.@device[0].name='br-lan'
network.@device[0].type='bridge'
network.@device[0].ports='lan1' 'lan2' 'lan3' 'lan4'
network.lan=interface
network.lan.device='br-lan'
network.lan.proto='static'
network.lan.ipaddr='192.168.1.50'
network.lan.netmask='255.255.255.0'
network.lan.ip6assign='60'
network.lan.dns='192.168.1.6'
network.lan.gateway='192.168.1.1'
network.wan=interface
network.wan.device='wan'
network.wan.proto='dhcp'
network.wan6=interface
network.wan6.device='wan'
network.wan6.proto='dhcpv6'
```

**uci show dhcp**

```
dhcp.@dnsmasq[0]=dnsmasq
dhcp.@dnsmasq[0].domainneeded='1'
dhcp.@dnsmasq[0].boguspriv='1'
dhcp.@dnsmasq[0].filterwin2k='0'
dhcp.@dnsmasq[0].localise_queries='1'
dhcp.@dnsmasq[0].rebind_protection='1'
dhcp.@dnsmasq[0].rebind_localhost='1'
dhcp.@dnsmasq[0].local='/lan/'
dhcp.@dnsmasq[0].domain='lan'
dhcp.@dnsmasq[0].expandhosts='1'
dhcp.@dnsmasq[0].nonegcache='0'
dhcp.@dnsmasq[0].cachesize='1000'
dhcp.@dnsmasq[0].readethers='1'
dhcp.@dnsmasq[0].leasefile='/tmp/dhcp.leases'
dhcp.@dnsmasq[0].resolvfile='/tmp/resolv.conf.d/resolv.conf.auto'
dhcp.@dnsmasq[0].nonwildcard='1'
dhcp.@dnsmasq[0].localservice='1'
dhcp.@dnsmasq[0].ednspacket_max='1232'
dhcp.@dnsmasq[0].filter_aaaa='0'
dhcp.@dnsmasq[0].filter_a='0'
dhcp.lan=dhcp
dhcp.lan.interface='lan'
dhcp.lan.start='100'
dhcp.lan.limit='150'
dhcp.lan.leasetime='12h'
dhcp.lan.dhcpv4='server'
dhcp.lan.ignore='1'
dhcp.wan=dhcp
dhcp.wan.interface='wan'
dhcp.wan.ignore='1'
dhcp.odhcpd=odhcpd
dhcp.odhcpd.maindhcp='0'
dhcp.odhcpd.leasefile='/tmp/hosts/odhcpd'
dhcp.odhcpd.leasetrigger='/usr/sbin/odhcpd-update'
dhcp.odhcpd.loglevel='4'
```

**Comment by bluewavenet** [comment]
@etiago 
Nothing immediately springs out in your configs as being wrong, but bare in mind that vlans will not be propagated over wireless links without some form of tunnelling. The upcoming mesh11sd will support guest/iot networks by default without needing vlans (but it will also have the capability to support vlans if so desired), but that is another story not effecting this issue.

As you do not seem to be doing anything special on your "dumb ap", I would suggest starting from scratch using the mesh11sd rapid deployment install.
This does not use the "dumb ap" approach at all. It does all the config for you in a well tried and tested way. It will give you a common flash image for all your mesh nodes - assuming they are all the same hardware.

https://openwrt.org/docs/guide-user/network/wifi/mesh/rapiddeployment

**If you want to try this, let me know and I will show you exactly what to do for your hardware as we follow the documentation link above.**

**Comment by etiago** [comment]
@bluewavenet

> Nothing immediately springs out in your configs as being wrong, but bare in mind that vlans will not be propagated over wireless links without some form of tunnelling. The upcoming mesh11sd will support guest/iot networks by default without needing vlans (but it will also have the capability to support vlans if so desired), but that is another story not effecting this issue.

I think the vlans shouldn't be playing a role here, as only the router should have any understanding and knowledge of the vlans. The port that connects the router to the AP is untagged so I'm expecting all packets going out to the AP will not have any vlan tags on them (this is how I would like it at least anyway).

> As you do not seem to be doing anything special on your "dumb ap", I would suggest starting from scratch using the mesh11sd rapid deployment install.

I have tried that in the past, but I was slightly confused as to what this:
```
uci set network.lan.ipaddr='192.168.9.1'
```
should be set to in the startup script.

This looks like it wants to be setting the IP address of the LAN interface of my OpenWRT AP, which in my case should then be 192.168.1.50 (that's the static IP I want on my AP). I have tried that, but then, even with just a single mesh node, I was unable to reach it altogether and had to hard reset it.

And then since I couldn't get this to work, and I was having issues even without attempting to form a mesh, I temporarily gave up on the meshing until I could figure out why I was getting the other issues...

I think I'm going to try doing the same setup but with a D-Link DAP-X1860, just in case it's something dodgy with this router... The D-Link has a similar chipset and radio though, so if it's an issue with the drivers, I think I'll get exactly the same 😢

**Comment by etiago** [comment]
Update:

I've tried the same exact setup with the D-Link DAP-X1860. To recap that's:

- mesh11sd, kmod-nft-bridge, wpad-openssl (minus wpad-basic-mbedtls).
- mesh11sd left on default configs.
- Each radio with a single SSID.

And lo and behold, just a couple of minutes after, I have some devices on wifi that can't get a DHCP response and end up self-configuring with `169.x` IP addresses.

I'm going to do one more test tonight where I just install `kmod-nft-bridge` but no `mesh11sd`. Considering `mesh11sd` just appears to be sitting idly since I haven't given it any config, I'm more inclined to believe the issue is being caused by the `kmod-nft-bridge`... and if I just install that one and the issue is still there, that should narrow the problem significantly...

**Comment by bluewavenet** [comment]
@etiago 
It may perhaps be some issue/interaction with the Microtik router.... mmmm.

Rather than mess around I think you should try the rapid deployment method.
It is tried and tested and used in large rollouts with no such problems.

For clarification, Rapid Deployment creates a single image per hardware type for rollout to any number of mesh devices and is 100% auto configuring.
There are two types of basic mesh nodes:
 1. Portal - has a wan connection to an upstream Internet feed and supports its own ipv4 subnet, routing between upstream and the mesh backhaul.
 2. Peer - joins the mesh backhaul and does not know or care about ipv4 networking as it bridges everything. 
 
There are numerous sub types also, but more on that later.

`uci set network.lan.ipaddr='192.168.9.1'`

This is the ipv4 gateway address that will be used by the mesh network and will be the ip address of the portal node.
As any node could be a portal, this address is embedded into the rapid deployment image. It is not the ip address of peer nodes.
Peer nodes will be allocated an ip address by dhcp-v4 if found on the network.

Mesh11sd provides tools to ssh to all nodes even without knowing an ipv4 address, so no need for a static ipv4 address (but if you really want this we can do it later)

I realise your intention is to run without a portal node and have user devices get an ipv4 address from the Microtik, but bear with me, we can do that after we check it all works correctly in the default mode of having a portal router connected to your Microtik.

One step at a time.... Using a known good config provided by Rapid Deployment.
If we have the same problem with this then most likely it is the chipset.
If we don't, then it is most likely the Microtik dhcp throwing a hissy fit.



I will post next the steps for rapid deployment using the Firmware Selector.

**Comment by etiago** [comment]
@bluewavenet sure, happy to give it a try 👍

**Comment by bluewavenet** [comment]
@etiago 
Follow this link:
https://openwrt.org/docs/guide-user/network/wifi/mesh/rapiddeployment#using_the_firmware_selector_to_create_a_custom_image

Select the Totolink X5000R.
The installed packages list should be as in this screenshot:

![image](https://github.com/user-attachments/assets/bade4858-3005-4007-ba02-4027788c1a51)

The "Script to run on first boot (uci-defaults)" box should be exactly the same as in the link.

Finally reflash the Totolink.

When done, connect the "wan" port of the Totolink to a "lan" port on the Microtik and power it up.
Mesh11sd will detect this and configure as a portal.

You should see the ssid "OpenWrt-xxxxx" once it has finished booting up.
Connect a device to it and you should get an ipv4 address in the 192.168.9.x range, with Internet access.

If this works, we can add a Totolink peer node using the same flash image.

If that works we can look at customising where necessary.

**Comment by etiago** [comment]
@bluewavenet I've tried this with a Totolink + a D-link (I ran out of usable Totolinks) and it all works fine.

As you requested, I plugged the Totolink's WAN port to the Mikrotik, and left the D-Link (flashed with a firmware using the exact same config) and this D-Link meshed with the Totolink just fine.

As you can see here, there's 1 active peer:
```
root@meshnode-4a20:~# mesh11sd status

{
  "setup":{
    "version":"4.1.1",
    "enabled":"1",
    "procd_status":"running",
    "portal_detect":"1",
    "portal_detect_threshold":"0",
    "portal_channel":"default",
    "channel_tracking_checkinterval":"30",
    "mesh_basename":"m-11s-",
    "auto_config":"1",
    "auto_mesh_network":"lan",
    "auto_mesh_band":"2g40",
    "auto_mesh_id":"ad96569faef7a1295f33a0790c8782",
    "mesh_gate_enable":"1",
    "mesh_leechmode_enable":"0",
    "mesh_gate_encryption":"3",
    "txpower":"20",
    "mesh_path_cost":"10",
    "mesh_path_stabilisation":"1",
    "checkinterval":"10",
    "interface_timeout":"10",
    "ssid_suffix_enable":"1",
    "debuglevel":"1"
  },
  "interfaces":{
    "m-11s-0":{
      "mesh_retry_timeout":"100",
      "mesh_confirm_timeout":"100",
      "mesh_holding_timeout":"100",
      "mesh_max_peer_links":"16",
      "mesh_max_retries":"3",
      "mesh_ttl":"31",
      "mesh_element_ttl":"31",
      "mesh_auto_open_plinks":"0",
      "mesh_hwmp_max_preq_retries":"4",
      "mesh_path_refresh_time":"1000",
      "mesh_min_discovery_timeout":"100",
      "mesh_hwmp_active_path_timeout":"5000",
      "mesh_hwmp_preq_min_interval":"10",
      "mesh_hwmp_net_diameter_traversal_time":"50",
      "mesh_hwmp_rootmode":"4",
      "mesh_hwmp_rann_interval":"5000",
      "mesh_gate_announcements":"1",
      "mesh_fwding":"1",
      "mesh_sync_offset_max_neighor":"50",
      "mesh_rssi_threshold":"-65",
      "mesh_hwmp_active_path_to_root_timeout":"6000",
      "mesh_hwmp_root_interval":"5000",
      "mesh_hwmp_confirmation_interval":"2000",
      "mesh_power_mode":"active",
      "mesh_awake_window":"10",
      "mesh_plink_timeout":"10",
      "mesh_connected_to_gate":"1",
      "mesh_nolearn":"0",
      "mesh_connected_to_as":"1",
      "mesh_id":"ad96569faef7a1295f33a0790c8782",
      "device":"radio0",
      "channel":"1",
      "tx_packets":"1673",
      "tx_bytes":"654159",
      "rx_packets":"1180",
      "rx_bytes":"271868",
      "this_node":"5c:92:5e:c4:4a:20",
      "active_peers":"1",
      "peers":{
        "66:29:43:a4:85:58":{
          "next_hop":"66:29:43:a4:85:58",
          "hop_count":"1",
          "path_change_count":"1",
          "metric":"33"
        }
      },
      "active_stations":"1",
      "stations":{
        "3c:21:9c:fe:7d:97":{
          "proxy_node":"66:29:43:a4:85:58"
        }
      }
    }
  }
}
```

Where to go from here?

Some thoughts on what I'm still missing:
- I'd like the wireless devices to be on the same LAN as the rest of the network. This setup puts them in a separate 192.168.9.0/24 lan.
- I'd like DHCP reservations to be managed by the existing DHCP server in the network.

If we can get these two working, I'd be a happy camper 😄

**Comment by bluewavenet** [comment]
@etiago 

> 
>  * I'd like the wireless devices to be on the same LAN as the rest of the network. This setup puts them in a separate 192.168.9.0/24 lan.
>  * I'd like DHCP reservations to be managed by the existing DHCP server in the network.

Move the ethernet cable from the Totolink wan port to a lan port. It should reconfigure over a few minutes, but to make it quick, power cycle both mesh nodes.

The Totolink will configure itself as a peer, bridging to the Microtik.

Now we will see if the Microtik dhcp was responsible for your problem.......

Note: Do not use Luci on meshnodes in auto_config mode as it is incompatible because Luci creates static configs, whereas mesh11sd manages dynamic configs that Luci cannot understand.
We can go into this and what can be done about it later if you really need Luci, although in my opinion it is not needed for a meshnode anyway.

**Comment by etiago** [comment]
@bluewavenet that did it! 😄 

The Totolink got a DHCP allocation from the Mikrotik and I was able to join its wireless and get DHCP allocations on the clients as well.

I'll need to run it for a bit longer (probably tonight) as the DHCP issue I was seeing needed to run for a little longer and I haven't yet had a chance to test it.

I think with this setup I won't be able to easily have my mesh nodes using static IPs, right? It's fine if that's the case. Would be nicer to have them with static IPs because sometimes I'll log into them to do general maintenance (updates, etc) and having them have predictable IPs outside the DHCP range is nice. But not a dealbreaker anyway.

When you say "don't use Luci"... Can I at least use it to configure the wireless for the AP? Or should I be doing something different for that config?

**Comment by bluewavenet** [comment]
@etiago 

> I won't be able to easily have my mesh nodes using static IPs

No but you should be able to configure static allocations on the dhcp server on the Microtik.

However mesh11sd gives you remote connection options for nodes (uses ssh on ipv6), specifically to enable remote maintenance.
ie see: `mesh11sd connect`

The default ssid is taken from the static wireless config.
Setting the ap wireless stuff, eg encryption details etc is done in the mesh11sd config.
These are simple to do but is most convenient if done in the image generation. 

See the full list of config options here:
https://github.com/openNDS/mesh11sd#default-configuration-file-etcconfigmesh11sd

A custom ssid is also best set in the image, but this depends on the details of the hardware layout.

A mesh11sd option to set this, overriding the wireless config, will be available in the upcoming v5.0.0 (release in the next week or two). 

If you want to view the **_dynamic_** config that mesh11sd is using, look at the outputs of:

`uci show mesh11sd`
`uci show wireless`
`uci show network`
`uci show dhcp`

**Note**: these will _**NOT**_ be the same as looking at `/etc/config/wireless` etc. as `/etc/config` holds the _**static**_ values.

**If you would like to try changing the ssid manually,** post the output of:
`uci show wireless` and I will show you what to do.... This will also determine what to put in the "Script to run on first boot (uci-defaults)" box on the Firmware Selector page to burn it into your image(s).

**Comment by bluewavenet** [comment]
@etiago 

> Do you know of any good guides on how to bake the custom SSID onto the image itself?

I did say I would help you with this.

Before we start, go back to your image before you added the extra stuff.

A note first, I see you have some 802.11r config lines.
Do you really need this?
Generally you will not and in fact it will cause more problems.

Mesh11sd, by default, adds a suffix to the configured ssid to ensure each access point has a unique ssid.
This is deliberate to make roaming much more efficient in most cases. Yes you have to take your phone or whatever to each area and connect - _**for the first time only**_ -. Thereafter it will be automatic.

_**We can discuss this later. But for now lets start simple.**_

**Note**: `wifinet0` and `wifinet1` will not exist. So your additional uci commands will be ignored.

Right then, here is a list of things to do.

 1. Get back to the original image we got working.
 2. SSH into a node
 3. Run `service mesh11sd stop`
 4. Run `uci revert wireless`
 5. Run `uci show wireless`
    From this output, look for lines containing `ssid=`
   It will look something like:
   `wireless.default_radio0.ssid='OpenWrt'` 
   and
    `wireless.default_radio0.ssid='OpenWrt'` 
 6. Change the 'OpenWrt' part to 'Etiago' or whatever you want
 7. Run commands like:
  `uci set wireless.default_radio0.ssid='Etiago'` 
  `uci set wireless.default_radio0.ssid='Etiago'` 
 8. Commit the changes with `uci commit wireless`
 9. Restart mesh11sd with `service mesh11sd restart; exit`
 Note the exit command, there to prevent possible loss of connection when the network restarts hanging up ssh.

Typically the new ssid(s) will look something like `Etiago-2g-xxxx` and `Etiago-5g-xxxx`

Once you see this is working you can build it into your flash image by adding just the two lines:
  `uci set wireless.default_radio0.ssid='Etiago'` 
  `uci set wireless.default_radio0.ssid='Etiago'` 

_**Note: This will not be necessary with mesh11sd v5.0.0 onwards as it will be built in to the mesh11sd config as an option.**_

**Comment by bluewavenet** [comment]
@etiago 

> On the 802.11r, I thought this made roaming between nodes faster 

Theoretically it could be faster but in practice it often is not. In addition, many client devices do not work well with it at all, sometimes as far as refusing to connect.

If you are going to be livestreaming while wandering around your house you might need something special. But usually it will be just fine. With or without 802.11r you might need to adjust transmit power on aps and adjust locations, so all in all it is not worth the effort in most cases.

**Comment by etiago** [comment]
@bluewavenet 

The instructions worked a treat.

I have a few more generic guidance type of questions:

1. In your experience, even if I don't use 802.11r, must I have different SSIDs for most clients to behave nicely when transitioning between nodes? The reason I ask is that I already have so many devices using my current SSID, and having to change all of them to use the new SSIDs with the MAC suffix is a bit inconvenient. I'm contemplating whether I can just set `ssid_suffix_enable` to 0 and still get decent client behaviour.
2. Best as I can tell, I can only disable the MAC suffix in the `mesh11sd` settings. Is it not possible to disable the addition of the `-2g` and `-5g` to the SSID? Again, same story as point 1. Then again, if this is not possible to disable, I might just bite the bullet and leave the MAC suffix enabled since I'd have to re-configure all my wireless clients anyway...

I always thought that having a single SSID across multiple APs and even across different frequency radios was the way to go, although what we're discussing here and this default behaviour is leading me to think that maybe I haven't been doing things the most optimal way...

**Comment by etiago** [comment]
Answered my own question on point 2. Looking through the mesh11sd code suggested that `ssid_suffix_enable` also controlled whether the `2g/5g` suffix was added. So I've now set that to `0` and I can have the exact ssid I had before 😄 

Now whether this is a good idea or not... that I still don't know...

**Comment by etiago** [comment]
@bluewavenet 

Unfortunately, looks like I'm right back where I started.

Left the network overnight with two nodes meshed (a Totolink and the D-Link). It was working for a few minutes last night before I went to bed.

When I arrived this morning I couldn't get a DHCP allocation.

Unplugged the remote node, leaving me just with the Totolink hardwired to the Mikrotik (with the DHCP server) and still, no DHCP reservation.

When I gave myself a static IP, I'm able to browse the Internet...

**Comment by bluewavenet** [comment]
@etiago 

> When I gave myself a static IP, I'm able to browse the Internet...

So from this we can pretty safely conclude the issue lies somewhere with the Microtik dhcp server. Possibly microtik is rejecting arp from the totolink connected clients... 

Lets verify it works reliably with the wan connection. Reconnect the ethernet so it goes from a microtik lan port to the totolink wan port, like we first tested it then power cycle.
If you give it a day or two to see if the problem re-occurs. If it does not happen with the wan connection (portal mode) then this will be strong evidence pointing at the problem. The new mesh11sd v5.0.0beta does have arp tuning options for edge cases so if we get to that stage it will be worth testing with it - but first things first!

**Comment by etiago** [comment]
@bluewavenet 

> So from this we can pretty safely conclude the issue lies somewhere with the Microtik dhcp server.

Based on what I've seen, I don't think I agree. Because:
1. This set up used to work with an older version of OpenWRT. Everything else being the same.
2. This set up works as long as I don't install `mesh11sd` and `kmod-nft-bridge`.

> Lets verify it works reliably with the wan connection. Reconnect the ethernet so it goes from a microtik lan port to the totolink wan port, like we first tested it then power cycle.

I'm fairly confident that this will work. I suspect `kmod-nft-bridge` maybe together with some default config somewhere is interfering with either the DHCP broadcast reaching Mikrotik, or with the response making it back... I expect setting it up with wan where DHCP runs directly on the mesh node should cause no issues.

I will try it anyway, but first I'm going to try just installing `kmod-nft-bridge` **but not mesh11sd**. I suspect this package on its own will cause the same behaviour.

**Comment by bluewavenet** [comment]
@etiago 

> I suspect this package on its own will cause the same behaviour.

I understand you might suspect this, but this has been tested and used in large rollouts with no problem.
`kmod-nft-bridge` is an nftables addon configured by mesh11sd to block mesh bridge loop storms from occurring.
If mesh11sd is not running/installed there will be no rules in the nft ruleset using the addons, _**so it is most definitely NOT the issue here**_ (unless you have installed something else that sets up layer 2 bridge filtering rules).

_**The arp issue I mentioned earlier does cause a problem with some dhcp services where client devices are behind the mesh layer 2 mac-routing running on meshnodes**_ - this is a known issue and enhancements to mesh11sd v5 have been added to mitigate the problem for non-portal ethernet uplinks.
I am pretty sure this is what you are seeing.

You can manually try the same thing as v5 does.

Assuming your mesh interface is `m-11s-0` and bridge interface is `br-lan` (you can use use `ip link` to check)

SSH to the Totolink with the ethernet link from lan port to the Microtik lan port, issue the following commands for testing purposes:

`echo 1 > /proc/sys/net/ipv4/conf/m-11s-0/proxy_arp_pvlan`
and
`echo 0 > /proc/sys/net/ipv4/conf/br-lan/proxy_arp`

These settings are volatile and will not survive a restart of mesh11sd, network, wireless or a reboot.

**Comment by etiago** [comment]
@bluewavenet 

You were right. `kmod-nft-bridge` in isolation had no impact at all. Been running happily the whole day without issue.

I will manually run the commands you suggested and leave it running the whole night.

**Comment by etiago** [comment]
@bluewavenet 

I don't want to jinx it but that seems to have done it 😄 although with a caveat.

I left it running with 2 nodes (Totolink + D-link) and applied those settings on the Totolink. When I tried it this morning, **the remote node wasn't giving me Internet access**.

When I unpluigged this mesh node and connected to the Totolink's SSID I did get a DHCP allocation. So I'm pretty confident this has fixed the Totolink at least.

But now I'm wondering:

--
#### Had the remote node perhaps disconnected due to hitting the `mesh_rssi_threshold`? I've plugged it back it and will do some more digging when it happens again (currently it's working fine).

Doing a few measurements of the signal strength on the mesh node just now, it seems to hover around 58 to 61dBm... and I'm using the default threshold of 65dBm....

--

#### Do I perhaps need to set `proxy_arp` on the mesh nodes' `br-lan` as well? (`proxy_arp_pvlan` was already set to 1 on the Totolink...)

--

And finally, I suspect it's not going to be easy to make this a persistent change in v4? From what you said it sounds like I'd need to run this any time `mesh11sd` (and/or the AP) is restarted...

I suppose I could always brute force it and have a cron running every 5 seconds that sets the `proxy_arp` to 0... 😄

**Comment by bluewavenet** [comment]
@etiago 

>  I suspect it's not going to be easy to make this a persistent change in v4? 

Correct, but you can manually upgrade to mesh11sd v5.0.0 beta for testing.

**With regard to testing,** I dug out an ancient and retired Microtik RB750 and set it up with a single meshnode connected just like yours, on mesh11sd v4.1.1. I set a 5 minute lease on routeros. Sure enough, I replicate your problem when the lease expired.

Replacing the RB750 on routeros with a gl-inet mt6000 on openwrt 25.05.5, same lease time, and no problems.

I upgraded to mesh11sd 5.0.0beta and back to the Microtik and the problem is fixed.

There is more to it than just those two test commands I suggested, obviously, but it was worth a try. The issue seems to be that the 802.11s mac-routing protocol makes the Microtik reject the dhcp renew because it thinks the arp reply is coming from the meshnode and not the client because the original arp table entry has expired.

_**If you are up for giving it a try, I'll package v5.0.0beta ipk for you for a manual install on your image.**_

**Comment by bluewavenet** [comment]
@etiago 

You can download here:
https://github.com/openNDS/mesh11sd/blob/master/community/testing/mesh11sd_5.0.0-beta-r1_all.ipk

Copy it to /tmp on the Totolink.
Then do:
`opkg install /tmp/mesh11sd_5.0.0-beta-r1_all.ipk; exit`

Note the `exit`. Mesh11sd will restart the network and this could hang your ssh session and you might have to wait 5 minutes for the session to time out before you can get back in. The `exit` prevents this and saves time.

**Comment by etiago** [comment]
> The issue seems to be that the 802.11s mac-routing protocol makes the Microtik reject the dhcp renew because it thinks the arp reply is coming from the meshnode and not the client because the original arp table entry has expired.

Most intriguing!

Wonder whether this is actual buggy behaviour from the Mikrotik or just it being overzealous...

> You can download here:

Thanks! I'll install this on both the Totolink and the D-link for good measure and leave it running overnight. 🤞

**Comment by bluewavenet** [comment]
@etiago 
It will preserve the current mesh11sd config. Nothing extra needed at this stage. Lots of new things to tweak if necessary.

**Comment by etiago** [comment]
@bluewavenet it's been running since last night for a good few hours, with a bunch of clients coming in and out, and it's been solid with two nodes👍 

Will add a couple more nodes to the mesh and hopefully that'll be me done 😄 

Many many thanks for all your help throughout this!

**Comment by bluewavenet** [comment]
@etiago 

> Will add a couple more nodes to the mesh 

Sure, but then you may need to consider location of the nodes.

Large overlaps of coverage can cause instability in the mesh paths as the wireless frequencies in use (microwaves) can cause multipath interference that changes not only spatially but also as a function of time.
For example, the "best path" might change as you walk around, open doors etc. This has little or no effect at layer 2, but can, for example, cause tcp links to break. But all this is another story we can get into if required.

The default for v5 onwards is "leechmode". A leech is a peer that connects to a backhaul but does not contribute actively to the backhaul. This means that multiple hops will be discouraged, making the mesh a "single hop  spoke" style network.
This is the best starting configuration.
If you need multiple mesh nodes with multiple hops then leechmode would be turned off on some or all nodes depending on many things. Other options are auto stabilisation, transmit power and rssi threshold.

But as I said, we can deal with this as and when we need to.

**Comment by etiago** [comment]
I think I'm all sorted now 😄 

I've gone as far as setting up a local build environment using the image builder and built images for my devices with v5 so now it's literally flash and forget.

I'd never played around with building images with built-in config but this is so much more convenient than having to manually fiddle with config in Luci every time...! 😄 

Happy to close this issue if you are.

**Comment by bluewavenet** [comment]
@etiago 
Any problems with more than two meshnodes? 

>  it's literally flash and forget.

That is the whole idea ;-)

> built images for my devices with v5

The beta you have is already quite a few commits behind as we are doing pre-release testing with certain parts disabled, and enabling those parts and committing to the beta. 

New functionality that will be in the release will include:
 * Guest/Iot wireless networking over the mesh backhaul, without needing vlans.
 * Opportunistic wireless encryption (owe), particularly desirable for guest networks), with owe transition to support older non-owe devices.
 * Central Access Point data collection and monitoring - where you can get info about client devices, where they are connected, how much data they are using etc.

> Happy to close this issue if you are.

Yes, seems a good point to do that. Go ahead when you are ready.

If you need help tuning with more nodes, open a new issue.

**Comment by etiago** [comment]
@bluewavenet 

> Any problems with more than two meshnodes?

I've just added another one (a D-Link X1860) and it's running without issues.

One thing I did notice is that the `path_change_count` is showing relatively high numbers and I don't know whether that's normal.

Referring to the values below which I've anonymised, `<A>` is the Totolink with the hardwire. `<B>` and `<C>` are the remote nodes.

From `<A>`:
```
      "<B>":{
          "next_hop":"<B>",
          "hop_count":"1",
          "path_change_count":"51",
          "metric":"36"
        },
        "<C>":{
          "next_hop":"<C>",
          "hop_count":"1",
          "path_change_count":"77",
          "metric":"27"
        }
```

From `<B>`:
```
        "<C>":{
          "next_hop":"<C>",
          "hop_count":"1",
          "path_change_count":"45",
          "metric":"29"
        },
        "<A>":{
          "next_hop":"<A>,
          "hop_count":"1",
          "path_change_count":"1227",
          "metric":"209"
        }
```

From `<C>`:
```
        "<B>":{
          "next_hop":"<B>",
          "hop_count":"1",
          "path_change_count":"29",
          "metric":"41"
        },
        "<A>":{
          "next_hop":"<A>",
          "hop_count":"1",
          "path_change_count":"293",
          "metric":"24"
        }
```

In particular, the connection between the remote nodes and the one node with the hard link (`<A>`) shows a pretty high `path_change_count`... but maybe that's entirely normal.

At least I haven't experienced any issues whilst using the network... I'll probably just let it be...

**Comment by bluewavenet** [comment]
@etiago 
If the path change count is increasing significantly over time it means that the nodes can see each other ok but you are getting a multipath interference pattern that is varying with time (as I tried to describe above).

You can monitor what is happening with:

```
mesh11sd debuglevel 3
logread -e mesh11sd -f 
```

You can also monitor the instantaneous connection values with:
`mesh11sd stations`

where the "signal" and "signal average" values are most interesting.

With regard to the multipath interference, I do not know your level of knowledge of the physics of wave propagation, but use the analogy of waves on an enclosed body of water if you throw a handful of stones into it. Waves will expand and mix with each other giving a chaotic looking pattern. I hope this helps ;-)

If you are not having issues with connections then fine. But if you are then there are lots of things that can be done to tune the system.

The very simplest is to move the nodes a little (even a few centimeters can have a big effect).
Monitor path changes over time.

**Comment by etiago** [comment]
@bluewavenet 

Sadly ended up with new issues today. The network was stable the whole day, I was in and out of video calls and always going strong with no issues.

However, at some point in the afternoon I lost Internet connectivity. Tried turning off the mesh nodes one by one until just the last AP and even with just the one AP wired to the Mikrotik, I wasn't able to even associate with the AP.

This seemed like a deeper issue with the radio or something as it wouldn't even connect at all. On an iPad it kept saying it couldn't connect. Whereas when there are DHCP issues, it normally connects but just says "No Internet".

I was in the middle of cooking dinner so didn't have time to spend debugging it and just replaced the one AP with another one I have without mesh11sd...

Fully appreciate I'm running a beta version of mesh11sd though 😄 I think I'd better wait for the final v5 release...

**Comment by bluewavenet** [comment]
@etiago 
One of my first comments in this thread was:

> This device is a peculiar one. It has a "last generation" mt7621 SoC but has a "next generation" mt7915 wireless chip.
> 
> There were known issues with the mt79xx series chips but this was fixed prior to the release of 23.05.5. But as this device is unusual, perhaps it was not fixed on this one.... I don't know.
> 
> The issue was a buffer overflow in the wireless drivers causing either a crash or for layer 3 traffic being dropped.

I do not know if the totolink is effected or not.

It happens when a radio hardware has to support more than one interface on the physical radio.
Unfortunately it looks like the fix was not backported to 23.05.xx 
I _**think**_ it is in the as yet unreleased 24.10.00. It is however in master/snapshot.

The problem happens in high traffic, mostly, resulting in a slow degradation in wireless performance then a sudden stop for all layer 3 traffic. Only a reboot seems to fix it.....

Mesh11sd v5 does have built in mitigation for circumstances like this - a new option set `reboot_on_error` and `portal_detect_threshold` but I don't think it is enabled on your beta.
What does `mesh11sd status` say?

> Tried turning off the mesh nodes one by one until just the last AP and even with just the one AP wired to the Mikrotik, I wasn't able to even associate with the AP.

So the problem was with this totolink I think. Try swapping it with the dlink.

**Comment by etiago** [comment]
> I do not know if the totolink is effected or not.

Thinking back, I do every now and then experience strange behaviour with the Totolink. Previously it was just the 5G radio that would stop working entirely. The SSID was advertised but I was unable to connect. Restarting just the radio (in Luci, I don't entirely know which parts it was restarting) would fix it for another while.

I bought 3 of these Totolinks thinking I was buying a bargain and it turns out I bought not more than an ugly paperweight 😆 

> So the problem was with this totolink I think. Try swapping it with the dlink.

The D-Link is a bit awkward for me to use as the wired node because it's one of those that plugs directly to a socket and I don't have one easily accessible in that location...

I've set everything up back the way it was before and if it happens again I'll try just restarting the Totolink. It sounds like that in itself should fix it, without needing to restart anything else... and if that's the issue, I might be able to source a GL-MT3000 to replace the wired Totolink in a more permanent fashion...

**Comment by bluewavenet** [comment]
@etiago 

> I might be able to source a GL-MT3000 

It too has an mt79 chipset and most definitely has the same issue. I have a number of them. The "reboot_on_error" option originates with debugging the mt3000.

I have a couple on a recent snapshot and so far they seem to be ok......... I am not yet fully convinced, but it does seem better.

I also have a number of mt6000s. Now these are something else all together "A quantum leap" would be a good phrase to use. More expensive and a lot bigger, but performance wise well worth it.

**Comment by etiago** [comment]
@bluewavenet 

> I also have a number of mt6000s. Now these are something else all together "A quantum leap" would be a good phrase to use. More expensive and a lot bigger, but performance wise well worth it.

This page: https://openwrt.org/toh/gl.inet/gl-mt6000 suggests the MT6000 also has an mt79 chipset... is that page wrong? Or is it specifically the mt7981 that has the issue?

I might build a snapshot firmware to see if it makes the issue go away on the Totolink then... 👍

**Comment by etiago** [comment]
@bluewavenet 

> EDIT: Set that up. There will be a "latest" beta this afternoon so will build the .apk after that.

Sounds good.

I've been playing with my setup a bit more. Moved some cables around so that I could use the DLink X1860 as the wired node and temporarily ran with a 2 node mesh of wired D-Link to remote D-Link.

I thought even though the DLink has the same chip, it could maybe be an improvement on wireless performance since the device page claims the DLink can do 2T2R, so at least for the backhaul sounded appealing - particularly since I'd have two 2T2R capable APs talking to each other.

Unfortunately it didn't seem to make much of a difference. The backhaul was still hitting around the same ~120Mbps which is pretty much what I can get with the Totolink as the wired node and the DLinks as remote nodes...

Maybe the DLink doesn't really support 2T2R or perhaps it's not actually enabled on the firmware... I don't really know how to check this.

Another thing I just remembered is that I'm not setting the country code on the radios in my firmwares... I'm wondering whether this is having an impact by defaulting to settings that are more conservative than necessary. Will try setting that and see whether it makes a difference at all.

**Comment by bluewavenet** [comment]
@etiago 

> the remote nodes don't seem to be able to mesh with it...

They will eventually, but it could take some time.
The reason for this is that you are running without a mesh portal, ie an ethernet connection from **wan**.
A portal node will proactively look for remote nodes, whereas non portals, by default will just sit and wait for quite a while before sending a "here I am" type of message (the simplest mode of the HWMP mac-routing protocol).

You can configure the ethernet connected (lan) node to emulate a portal, but it is much more efficient to have an actual portal....

Other than not originally planning to have a separate ipv4 subnet for the mesh backhaul, is there any reason not to do so?
Any other devices currently connected to a Microtik lan port could instead be connected to a portal lan port... Just saying...

**Comment by etiago** [comment]
@bluewavenet 

> Other than not originally planning to have a separate ipv4 subnet for the mesh backhaul, is there any reason not to do so?
> Any other devices currently connected to a Microtik lan port could instead be connected to a portal lan port... Just saying...

Doesn't this add more strain to the AP's CPU by introducing NAT to the mix?

Currently I expect that any wireless traffic gets switched at layer 2 since wlan and lan are bridged, but operating in portal mode with a separate subnet I think then means any wireless traffic needs to go through NAT (with the extra load of layer 3 routing, maintaining NAT pinholes, etc).

**Comment by bluewavenet** [comment]
@etiago 

> Doesn't this add more strain to the AP's CPU by introducing NAT to the mix?

I do understand where you are coming from, but unless you are an avid gamer after every microsecond of latency, the effect is insignificant. If you were an avid gamer, you would have 10Gb/s ethernet everywhere anyway.
When you tried it in this mode earlier in this thread, did you notice a problem? I very much suspect not.

**Comment by bluewavenet** [comment]
@etiago 

> I bit the bullet and snatched a MT6000 on a black friday deal...

Lol ;-)

> a really capable device that could entirely replace my Mikrotik and plug directly to my FTTP modem.

Yes! And it would also be running as a  "portal".

> I really have a need to have a few devices with static IP allocations,

> Is it possible to have mesh11sd in auto-configure mode whilst manually defining static DHCP entries in OpenWRT?

Yes, make them static leases instead of static allocations, added to the dhcp config on the new portal mt6000 - and yes you could "bake" them into the mt6000 firmware.

>  I need to override the DNS server to point to a specific static IP in my network.

Again, bake it into the dhcp config in the mt6000 firmware

**Comment by bluewavenet** [comment]
@etiago 

_**REFLASHING THE MT6000**_

If you have not done it already, it is far better, and far safer to use the uboot-wbui and not the very confusing and wrong wiki instructions that will have you going round in circles unless you are lucky.

Instructions:

 1. Set your computer/laptop to ise ip address 192.168.1.1 static
 2. Connect ethernet from computer to any lan port of the mt6000
 3. With the mt56000 power lead unplugged, press and hold the reset button
 4. Plug in the power and wait for the led to finish flashing and go from blue to solid white
 5. On your computer, in the browser, go to http://192.168.1.1 and enter the uboot-webui screen
 6. Select the previously downloaded OpenWrt _**SYSUPGRADE**_ image.  _**Ignore ALL the other images.**_
 7. Upload and go, you should see "Update in progress"
 8. Wait a few minutes and the led will eventually go to solid white again. 
 9. Job done!

**Comment by etiago** [comment]
@bluewavenet

I've installed the MT6000 (initially just in LAN mode, with the Mikrotik still in the picture) and the irony is that it now looks like even without a mesh (just with the MT6000) I get really good performance throughout the whole house...

Bought this thing to improve my mesh, and it may have made the mesh unnecessary for my use 😭 

I'm still going to play around with adding some nodes to see how it behaves, in hopes that AP <-- far away--> AP has even better performance than AP --far away--> client, but the key takeaway is that this MT6000 truly is a beast. Puts the Totolink to shame...

**Comment by bluewavenet** [comment]
@etiago 

>  the key takeaway is that this MT6000 truly is a beast

Indeed it is. The 4 active antennas (effectively one stream/embedded radio per antenna) form a phased array giving dynamic beamforming on a per packet basis. The MT7986 supports 4 such streams.
You can see this in the outputs of:
`iw list | grep -A 8 "HE TX"`
and
`iw list | grep -A 8 "HE TX"`

The nextgen wireless chip supports 8 streams (the current max supported by `iwinfo`.
Coming in the not to distant future are 16, 32 and 64 stream chips...
The routers of the future will most likely be long and thin ;-)

> 
`I'm still going to play around with adding some nodes to see how it behaves, in hopes that AP <-- far away--> AP has even better performance than AP --far away--> client`

The performance will be improved, but bare in mind the client connection will be defined by the meshnode to meshnode backhaul so it might not be a very big improvement.

For that you need "mt6000" class mesh nodes. That is something  I have been testing.
I think I mentioned earlier, you can easily get greater than 1Gb/s between nodes on 2.4GHz HE40.
